src/iosrv/drpc_ras.c

```diff
@@ -0,0 +1,283 @@
+	}
+
+	if (rank) {
+		evt->rank = (uint32_t)*rank;
+		D_FPRINTF(stream, " rank: [%u]", *rank);
+	}
+
+	if (jid) {
+		evt->job_id = jid;
+		D_FPRINTF(stream, " jobid: [%s]", jid);
+	}
+
+	if (puuid && !uuid_is_null(*puuid)) {
+		D_ALLOC(evt->pool_uuid, DAOS_UUID_STR_SIZE);
+		if (!evt->pool_uuid)
+			D_GOTO(out_fail, rc = -DER_NOMEM);
+		D_ASPRINTF(evt->pool_uuid, DF_UUIDF, DP_UUID(*puuid));
+		D_FPRINTF(stream, " puuid: ["DF_UUIDF"]", DP_UUID(*puuid));
+	}
+
+	if (cuuid && !uuid_is_null(*cuuid)) {
+		D_ALLOC(evt->cont_uuid, DAOS_UUID_STR_SIZE);
+		if (!evt->cont_uuid)
+			D_GOTO(out_fail, rc = -DER_NOMEM);
+		D_ASPRINTF(evt->cont_uuid, DF_UUIDF, DP_UUID(*cuuid));
+		D_FPRINTF(stream, " cuuid: ["DF_UUIDF"]", DP_UUID(*cuuid));
+	}
+
+	if (oid) {
+		D_ALLOC(evt->obj_id, DAOS_RAS_STR_FIELD_SIZE);
+		if (!evt->obj_id)
+			D_GOTO(out_fail, rc = -DER_NOMEM);
+		D_ASPRINTF(evt->obj_id, DF_OID, DP_OID(*oid));
+		D_FPRINTF(stream, " oid: ["DF_OID"]", DP_OID(*oid));
+	}
+
+	if (cop) {
+		evt->ctl_op = cop;
+		D_FPRINTF(stream, " control_op: [%s]", cop);
+	}
+
+	fclose(stream);
+	D_DEBUG(DB_MGMT, "&&& RAS EVENT%s", buf);
+	free(buf);
+
+	return 0;
+out_fail:
+	free_ras(evt);
+	fclose(stream);
+	free(buf);
+
+	return rc;
+}
+
+static int
+send_ras(Shared__RASEvent *evt)
+{
+	Shared__ClusterEventReq	 req = SHARED__CLUSTER_EVENT_REQ__INIT;
+	uint8_t			*reqb;
+	size_t			 reqb_size;
+	Drpc__Call		*dreq;
+	Drpc__Response		*dresp;
+	int			 rc;
+
+	if (!evt) {
+		D_ERROR("null RAS event\n");
+		return -DER_INVAL;
+	}
+	req.event = evt;
+
+	reqb_size = shared__cluster_event_req__get_packed_size(&req);
+	D_ALLOC(reqb, reqb_size);
+	if (reqb == NULL)
+		D_GOTO(out, rc = -DER_NOMEM);
+
+	shared__cluster_event_req__pack(&req, reqb);
+	rc = drpc_call_create(dss_drpc_ctx, DRPC_MODULE_SRV,
+			      DRPC_METHOD_SRV_CLUSTER_EVENT, &dreq);
+	if (rc != 0) {
+		D_FREE(reqb);
+		goto out;
+	}
+
+	dreq->body.len = reqb_size;
+	dreq->body.data = reqb;
+
+	rc = drpc_call(dss_drpc_ctx, R_SYNC, dreq, &dresp);
+	if (rc != 0)
+		goto out_dreq;
+	if (dresp->status != DRPC__STATUS__SUCCESS) {
+		D_ERROR("received erroneous dRPC response: %d\n",
+			dresp->status);
+		rc = -DER_IO;
+	}
+
+	drpc_response_free(dresp);
+out_dreq:
+	/* This also frees reqb via dreq->body.data. */
+	drpc_call_free(dreq);
+out:
+	return rc;
+}
+
+void
+ds_notify_ras_event(ras_event_t id, char *msg, ras_type_t type, ras_sev_t sev,
+		    char *hid, d_rank_t *rank, char *jid, uuid_t *puuid,
+		    uuid_t *cuuid, daos_obj_id_t *oid, char *cop, char *data)
+{
+	Shared__RASEvent	 evt = SHARED__RASEVENT__INIT;
+	int		 rc;
+
+	if (dss_drpc_ctx == NULL) {
+		D_ERROR("dRPC not connected: "DF_RC"\n", DP_RC(-DER_UNINIT));
+		return;
+	}
+
+	/* use opaque blob oneof case for extended info for passthrough event */
+	evt.extended_info_case = SHARED__RASEVENT__EXTENDED_INFO_STR_INFO;
+	evt.str_info = data;
+
+	rc = init_ras(id, msg, type, sev, hid, rank, jid, puuid, cuuid, oid,
+		      cop, &evt);
```

Comment:
Please just have init_ras() initializing the evt and then have another call below to log the write event to debug (requirement from Cray to be able to synchronize things).

Reply:
ok